source/MaterialXCore/Document.h

```diff
@@ -98,6 +121,33 @@ class MX_CORE_API Document : public GraphElement
     /// nodes, and include both Input and Output elements.
     vector<PortElementPtr> getMatchingPorts(const string& nodeName) const;
 
+    /// Return the child element, if any, with the given document and data library
+    template <class T> shared_ptr<T> getChildOfType(const string& name) const
```

Comment:
This latest version looks excellent, @ashwinbhat!

For consistency, though, wouldn't `getChildOfType` and `getChildrenOfType` need to be virtual methods, with the `Document` class overriding the behavior of its base `Element` class?

In addition to providing callers with more consistent behavior, this would allow the `resolveNameReference` method below to be implemented without a `dataLibrary` argument, and its implementation would require only a single call to `getChildOfType`.

Reply:
@jstone-lucasfilm Just realized, since `getChildOfType` and `getChildrenOfType` are templated, they can't be `virtual`
https://christiandaley.github.io/2024/02/18/virtual-function-templates-with-stateful-metaprogramming-in-c++-20.html

Reply:
Ah, that's a great point!

As an alternative, what if we were to move the implementations for these templated methods into `Element.cpp`, where they would have access to the `Document` class?

That would allow us to use the approach proposed earlier in the thread, where the implementations of `getChildOfType` and `getChildrenOfType` would check whether the calling element is a `Document`, and handle the data library of the `Document` for that case.

That would have the same net effect, where the behavior of `getChildOfType` and `getChildrenOfType` would be consistent, regardless of the type of the calling pointer, and we could remove the last `dataLibrary` argument from `resolveNameReference`.

Reply:
I think moving them to the cpp will require us to explicitly instantiate each type for export. Right now we depend on the compiler and header includes so anyone who has custom derived types are automatically resolved. I think moving it the cpp might cause some additional integrations issues for some teams.
I think `resolveNameReference` is the only outlier, in future we might want to reconsider moving these.

Reply:
That's a good point, @ashwinbhat, though I'm torn about the API churn that would be created by adding a `dataLibrary` argument to `resolveNameReference`, and then very likely returning to our original API in a subsequent change.

When I'm working on significant changes like your current PR, one approach that I'll often use is to implement any required refactoring steps in a _separate branch_, where it's straightforward to validate that the refactor has no unexpected impact on other workflows.

Using that approach, either you or I could propose the move of `getChildOfType` and `getChildrenOfType` into `Element.cpp`, independent of your current work, and we could validate that this step is itself a safe one to make.

I'd be happy to help with this if it's too time-consuming to take such a step on your side, and if it works out well, we would merge it into your current PR, allowing it to be merged without any associated API churn.

Reply:
@jstone-lucasfilm any thoughts about add an API to document to resolve name reference? e.g. `resolveNameFromDataLibrary` 

Which would essentially do something this:
```
ElementPtr resolveNameFromDataLibrary(const string& name, ConstElementPtr parent = nullptr) const
{
    ElementPtr child = hasDataLibrary() ? getDataLibrary()->getChild(name) : nullptr;
    return child;
}
```

Reply:
That's a clever idea!  I can see that handling the case of `getChildOfType` very elegantly, but it's not clear how it would handle the case of `getChildrenOfType`, which needs to resolve more than a single name.

How are you imagining it would handle that second case?

Reply:
I was thinking about how to remove the use of datalibrary in `resolveNameReference` thus keeping all datalibrary usage in Document. I think the main caller is  `Node::getNodeDef` that needs to either `resolveNameReference` if `nodedef` attribute is in use otherwise falls to use `Document::getMatchingNodeDefs` 
https://github.com/AcademySoftwareFoundation/MaterialX/blob/b058d9cbba16d06ed858c53f39f2af7b9e41b550/source/MaterialXCore/Node.cpp#L71
We could introduce `Document::getMatchingNodeDef` as counterpart to `Document::getMatchingNodeDefs` 
Since `resolveNameReference` is not a public `Node::getNodeDef` could continue to use that, but use  `Document::getMatchingNodeDef` to resolve node definition from datalibrary.

Reply:
That sounds very reasonable, and let's see how things go using that approach!

I'm still open to moving `getChildOfType` and `getChildrenOfType` to `Element.cpp` if that's a useful adjunct to your change, and just let me know if that step might be needed here.